pkg/bloombuild/planner/planner.go

```diff
@@ -171,39 +183,106 @@ func (p *Planner) runOne(ctx context.Context) error {
+	work, err := p.loadTenantWork(ctx, tables)
 	if err != nil {
 		return fmt.Errorf("error loading work: %w", err)
 	}
 
+	// For deletion, we need to aggregate the results for each table and tenant tuple
+	// We cannot delete the returned tombstoned metas as soon as a task finishes since
+	// other tasks may still be using the now tombstoned metas
+	tasksResultForTableTenant := make(map[tableTenant]tenantTableTaskResults)
 	var totalTasks int
-	for _, w := range work {
-		logger := log.With(p.logger, "tenant", w.tenant, "table", w.table.Addr(), "ownership", w.ownershipRange.String())
 
-		gaps, err := p.findGapsForBounds(ctx, w.tenant, w.table, w.ownershipRange)
-		if err != nil {
-			level.Error(logger).Log("msg", "error finding gaps", "err", err)
-			continue
-		}
+	for table, tenants := range work {
+		for tenant, ownershipRanges := range tenants {
+			logger := log.With(p.logger, "tenant", tenant, "table", table.Addr())
+			tt := tableTenant{
+				tenant: tenant,
+				table:  table,
+			}
+			var tasks []*protos.Task
+
+			// Fetch source metas to be used in both build and cleanup of out-of-date metas+blooms
+			metas, err := p.bloomStore.FetchMetas(
+				ctx,
+				bloomshipper.MetaSearchParams{
+					TenantID: tenant,
+					Interval: bloomshipper.NewInterval(table.Bounds()),
+					Keyspace: v1.NewBounds(0, math.MaxUint64),
+				},
+			)
+			if err != nil {
+				level.Error(logger).Log("msg", "failed to get metas", "err", err)
+				continue
+			}
 
-		now := time.Now()
-		for _, gap := range gaps {
-			totalTasks++
+			for _, ownershipRange := range ownershipRanges {
+				logger := log.With(logger, "ownership", ownershipRange.String())
 
-			task := NewTask(
-				ctx, now,
-				protos.NewTask(w.table, w.tenant, w.ownershipRange, gap.tsdb, gap.gaps),
-			)
+				// Filter only the metas that overlap in the ownership range
+				metasInBounds := bloomshipper.FilterMetasOverlappingBounds(metas, ownershipRange)
+				level.Debug(logger).Log("msg", "found relevant metas", "metas", len(metasInBounds))
 
-			if err := p.enqueueTask(task); err != nil {
-				level.Error(logger).Log("msg", "error enqueuing task", "err", err)
-				continue
+				// Find gaps in the TSDBs for this tenant/table
+				gaps, err := p.findOutdatedGaps(ctx, tenant, table, ownershipRange, metasInBounds, logger)
+				if err != nil {
+					level.Error(logger).Log("msg", "failed to find outdated gaps", "err", err)
+					continue
+				}
+
+				for _, gap := range gaps {
+					tasks = append(tasks, protos.NewTask(table, tenant, ownershipRange, gap.tsdb, gap.gaps))
+				}
+			}
+
+			var enqueuedTasks int
+			resultsCh := make(chan *protos.TaskResult, len(tasks))
+
+			now := time.Now()
+			for _, task := range tasks {
+				queueTask := NewTask(ctx, now, task, resultsCh)
+
+				if err := p.enqueueTask(queueTask); err != nil {
+					level.Error(logger).Log("msg", "error enqueuing task", "err", err)
+					continue
+				}
+
+				totalTasks++
+				enqueuedTasks++
```

Comment:
The naming might be a bit confusing.
`totalTasks` is the number of all tasks of this current iteration
`enqueuedTasks` is the number of tasks for a tenant table